### PR TITLE
fix: works with safari 12.0.1

### PR DESCRIPTION
**Pull request by spike008t**
Hey!

Another PR to support Safari 12 (macOS 10.13.x and iOS 12.x).
Safari do not support anymore `RTCPeerConnection::addStream` and must use `RTCPeerConnection::addTrack`.

### CODE REVIEW

File: lib/RTCSession.js [hunk 1]
```diff
@@ -676,7 +676,17 @@ module.exports = class RTCSession extends EventEmitter
         this._localMediaStream = stream;
         if (stream)
         {
-          this._connection.addStream(stream);
+          if (this._connecting.addTrack)
```
Comment (ibc) [inline lib/RTCSession.js]: Have you tested this? I don't think `this._connecting.addTrack` is what you want...

File: lib/RTCSession.js [hunk 2]
```diff
@@ -676,7 +676,17 @@ module.exports = class RTCSession extends EventEmitter
...
+          {
+            stream.getTracks().forEach(track => {
+              this._connection.addTrack(track, stream);
+            });
+          }
+
+          if (this._connection.addStream)
+          {
+            this._connection.addStream(stream);
+          }
```
Comment (ibc) [inline lib/RTCSession.js]: Why should it call `pc.addStream()` after having called `pc.addTrack()` before?

**Comment by ibc** [comment]
In addition to above referenced issues, what about calls to `this._connection.getLocalStreams()` in RTCSession.js? That's also old stream-based API not supported in Safari.

File: lib/RTCSession.js (hunk 1 above)
Comment (spike008t) [reply lib/RTCSession.js]: arf right! I miss my copy/paste

Comment (spike008t) [reply lib/RTCSession.js]: fixed, sorry

**Comment by ibc** [comment]
Note that I did another comment in this conversation.

In addition: is this PR completely tested in Chrome, Firefox and Safari?

File: lib/RTCSession.js (hunk 2 above)
Comment (spike008t) [reply lib/RTCSession.js]: For people having old implementation when listening event `addStream`, they do not have to change the code.

Comment (ibc) [reply lib/RTCSession.js]: `pc.onaddstream` event is not fired when calling `pc.addStream()` but just when calling `pc.setRemoteDescription()` with a remote SDP offer/answer that produces remote streams and tracks.

**Comment by spike008t** [comment]
@ibc `RTCPeerConnection::getLocalStreams()` is not supported anymore on safari. 
capture d ecran 2018-12-04 a 22 56 50

That works under safari 12.0.1 on MacOS, Chrome 70.0.3538.110, i'll give a try for firefox.

File: lib/RTCSession.js (hunk 2 above)
Comment (spike008t) [reply lib/RTCSession.js]: Or should I check if `pc.addTrack()` exists, in case of that is not availble, fallback on `pc.addStream()` ?

**Comment by ibc** [comment]
> @ibc `RTCPeerConnection::getLocalStreams()` is not supported anymore on safari.

Right, but RTCSession.js uses it (https://github.com/versatica/JsSIP/blob/master/lib/RTCSession.js#L3318) and this PR does not change it :)

File: lib/RTCSession.js (hunk 2 above)
Comment (ibc) [reply lib/RTCSession.js]: Probably, yes.

**Comment by spike008t** [comment]
> > @ibc `RTCPeerConnection::getLocalStreams()` is not supported anymore on safari.
> 
> Right, but RTCSession.js uses it (https://github.com/versatica/JsSIP/blob/master/lib/RTCSession.js#L3318) and this PR does not change it :)

Oh! i didn't see it... I haven't test mute on my app... I'll give a try

**Comment by spike008t** [comment]
@ibc Ah okay I didn't understand that (must be a quite late at the moment). I'm gonna clean up the code and remove old code with streams

**Comment by ibc** [comment]
Hi. I prefer that @jmillan checks this PR. He is out for two weeks so let's wait a bit, please.

**Comment by spike008t** [comment]
@ibc no problem for checking the PR, in fact I do need to solve my app this week.

**Comment by davies147** [comment]
I though that the whole idea of `webrtc-adapter` was to resolve these compatibility issues so that individual applications such as JsSIP does not have to worry about it.

I understand that the webrtc-adapter dependency was removed from JsSIP's internals, but I believe that was to free up the app developer to use a version of their choice so JsSIP did not have to constantly keep updating.

Or am I missing something? Perhaps I should also be assuming that webrtc-adapter will ensure that new webrtc calls will be polyfilled on older browsers?

If it helps, I just tested un-patched JsSIP on Safari 12.0.1 with adapter.js 6.4.8 and it worked fine.

**Comment by ibc** [comment]
We removed `webrtc-adapter` because, due a bug in the `webrtc-adapter` NPM package` it was not being used at all by JsSIP (`require('webrtc-adapter')` did absolutely nothing).

Even without adapter, JsSIP WebRTC API usage must be modernized. We cannot use the super old stream-based API in 2019.

**Comment by ibc** [comment]
Oh @davies147, super clear now, and agreed :)

**Comment by nicketson** [comment]
I have created a similar PR request here: 

https://github.com/versatica/JsSIP/pull/553

We've been using this fix for some time without issues but haven't had the spare time to create a PR until now.

**Comment by jmillan** [comment]
Corresponding changes have already been applied by #553 

Thanks a lot!

**Pull request closed without merging**